### ISSUE TITLE
Add Rector to our code base

### ISSUE THREAD

**Issue by carlos-granados** [Behat/Behat#1506]
Once we have a coding standard tool in place, I think we could add [Rector](https://github.com/rectorphp/rector) to our code base. Having this tool available would allow us to easily:
- Update our code base to use the latest PHP capabilities once we have tied down the PHP versions that we support
- Increase our type coverage
- And in general keep a tidier and nicer code base

**Comment by ttomdewit** [comment]
I've had some bad luck with Rector many, many attempts ago. I'm all for adding it, that goes without saying, but I'd be very careful as to what's enabled and how it's ran in CI. I am not a fan of "big bang" PRs that change a lot of different things at the same time. My ideal would be to introduce the tool and start enabling rules/fixers one at a time. Once enabled it'll keep running in CI. I believe this is a more manageable and sensible approach.

**Comment by ttomdewit** [comment]
> I've had some bad luck with Rector many, many attempts ago. I'm all for adding it, that goes without saying, but I'd be very careful as to what's enabled and how it's ran in CI. I am not a fan of "big bang" PRs that change a lot of different things at the same time. My ideal would be to introduce the tool and start enabling rules/fixers one at a time. Once enabled it'll keep running in CI. I believe this is a more manageable and sensible approach.

Sorry, I missed https://github.com/Behat/Behat/issues/1306#issuecomment-2449323357 but it echos my ideas here as well.

**Comment by carlos-granados** [comment]
@ttomdewit yes, with Rector it makes sense to go slowly and not introduce a lot of changes at the same time

**Comment by acoulton** [comment]
@ttomdewit couldn't agree more. Rector (and ECS) are great tools but we've had problems on legacy projects where rules that should have been just code-style / internal have produced breaking changes because our original code was not in an expected state, so being able to review the changes sensibly is vital.

**Comment by stof** [comment]
Upgrading automatically to newer PHP features with Rector can also be risky as Rector does not consider the impact of the change on backward compatibility. So I would not add it as part of a CI step enforcing the code to match the Rector output. In my experience, the output of Rector requires code review on what it does (especially in the context of a package subject to semver).

Regarding increasing type coverage, I think this is more about adding static analysis in the project than about adding Rector (btw, Rector is generally risky to run if you don't have phpstan passing at a high level)

**Comment by rpkamp** [comment]
Agree with @stof. I wouldn't add Rector before PHPStan is also running at at least level 7. Ideally higher.

**Comment by acoulton** [comment]
I think it's OK so long as we are deliberate about which rules we're introducing & the diffs they are producing (which @carlos-granados has been so far)?

If a PR is carefully reviewed & the rule does not produce a BC break when first introduced, then I don't think there's a BC risk to enforcing it for subsequent new features & refactors etc?

**Comment by stof** [comment]
@acoulton depends on the rules indeed. Rules that are mostly coding styles are fine (but running php-cs-fixer or phpcs might enforce them with a tool that runs faster than rector). But any rector rule that relies on type information (which is the big benefit of implementing something as a rector rule rather than a php-cs-fixer fixer) would be unsafe if type information is not 100% reliable (which requires being on level 8 without ignored errors)
